### PR TITLE
[MIG] account_credit_control: Migration to 12.0

### PR DESCRIPTION
**Pull request by cubells**
Create README in oca fragments.

Standard migration.

cc @Tecnativa

### CODE REVIEW

File: account_credit_control/models/credit_control_line.py [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+                open_amount = move_line.amount_residual
+            cur_tolerance = tolerance.get(move_line.currency_id.id,
+                                          tolerance_base)
+            if check_tolerance and open_amount < cur_tolerance:
+                continue
+            vals = self._prepare_from_move_line(move_line,
+                                                level,
+                                                controlling_date,
+                                                open_amount)
+            line = self.create(vals)
```
Comment (astirpe) [inline account_credit_control/models/credit_control_line.py:233]: I think this version of the module could benefit from the batch create (`@api.model_create_multi`) mechanism. Just do not create the line here inside the `for move_line in lines` loop, but outside the loop, passing to the create method a `vals_list` variable with the list of the `vals` just created above.

File: account_credit_control/models/credit_control_line.py [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+        required=True,
+        index=True,
+        readonly=True,
+        states={'draft': [('readonly', False)]},
+    )
+    # maturity date of related move line we do not use
+    # a related field in order to
+    # allow manual changes
+    date_due = fields.Date(
+        string='Due date',
```
Comment (astirpe) [inline account_credit_control/models/credit_control_line.py:35]: No need of `string='Due date',`

File: account_credit_control/models/credit_control_line.py [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+        required=True,
+        readonly=True,
+        states={'draft': [('readonly', False)]},
+    )
+    date_entry = fields.Date(
+        string='Entry date',
+        related='move_line_id.date',
+        store=True,
+        readonly=True,
```
Comment (astirpe) [inline account_credit_control/models/credit_control_line.py:44]: In Odoo 12 the related fields are readonly by default. You probably meant `readonly=False`? Please also check the fields below.

File: account_credit_control/models/credit_control_run.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+        """
+        runs = self.search(
+            [('date', '>', controlling_date),
+             ('company_id', '=', self.env.user.company_id.id)],
+            order='date DESC',
+            limit=1,
+        )
+        if runs:
+            raise UserError(_('A run has already been executed more '
+                              'recently than %s') % (runs.date))
```
Comment (astirpe) [inline account_credit_control/models/credit_control_run.py:102]: No need of parenthesis around `runs.date`

File: account_credit_control/models/res_company.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+from odoo import fields, models
+
+
+class ResCompany(models.Model):
+    """ Add credit control parameters """
+    _inherit = 'res.company'
+
+    credit_control_tolerance = fields.Float(
+        default=0.1,
+    )
```
Comment (astirpe) [inline account_credit_control/models/res_company.py:13]: I think you need the `readonly=False` here and below.

File: account_credit_control/tests/test_account_invoice.py [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+from datetime import datetime
+from dateutil import relativedelta
+
+from odoo import fields
+from odoo.tests.common import TransactionCase
+from odoo.exceptions import UserError
+
+
+class TestAccountInvoice(TransactionCase):
+    post_install = True
```
Comment (astirpe) [inline account_credit_control/tests/test_account_invoice.py:12]: You could use the:

```python
from odoo.tests import tagged

@tagged('post_install', '-at_install')
```

File: account_credit_control/tests/test_account_invoice.py [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+        new_policy = self.env.ref(
+            'account_credit_control.credit_control_2_time')
+        days_30 = self.env.ref('account_credit_control.2_time_1')
+        wiz_obj = self.env['credit.control.policy.changer']
+        wiz_rec = wiz_obj.with_context(active_ids=invoice.ids).create({
+            'new_policy_id': new_policy.id,
+            'new_policy_level_id': days_30.id
+        })
+
+        # Verfiy the lines on wiz belongs to same invoice
```
Comment (astirpe) [inline account_credit_control/tests/test_account_invoice.py:278]: Typo: `Verfiy`

File: account_credit_control/views/credit_control_line.xml [hunk 8]
```diff
@@ -53,74 +53,74 @@
...
-                    <field name="policy_id"/>
-                    <separator />
-                    <field name="partner_id"/>
-                    <field name="partner_user_id"/>
-                    <field name="account_id"/>
-                    <field name="invoice_id"/>
-                    <field name="level"/>
-                    <field name="policy_level_id"/>
-                </group>
+<!--                <group>-->
```
Comment (pedrobaeza) [inline account_credit_control/views/credit_control_line.xml:56]: Why is all of this commented?

File: account_credit_control/views/credit_control_line.xml [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+            </form>
+        </field>
+    </record>
+
+    <record id="credit_control_line_search" model="ir.ui.view">
+        <field name="name">Credit Control Lines</field>
+        <field name="model">credit.control.line</field>
+        <field name="arch" type="xml">
+            <search string="Control Credit Lines">
+<!--                <group>-->
```
Comment (astirpe) [inline account_credit_control/views/credit_control_line.xml:56]: Work in progress?

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
Runbot warning:
```
WARNING openerp_test odoo.models: The model credit.control.policy.changer has no _description
```

and please squash several Weblate translation from the same person plus other superfluous commits like bump version, or even some adjacent of the same author within the same logic set

File: account_credit_control/views/res_config_settings_view.xml [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+               eval="[(4, ref('account_credit_control.group_account_credit_control_manager'))]"/>
+        <field name="arch" type="xml">
+            <xpath expr="//div[@id='invoicing_settings']" position="after">
+                <h2>Credit Control</h2>
+                <div class="row mt16 o_settings_container">
+                  <div class="row col-md-6 o_setting_box" id="credit_control_policy"
+                    title="This allows credit control team to chase overdue payments with different means">
+                    <div class="o_setting_left_pane"/>
+                    <div class="o_setting_right_pane">
+                        <label string="Default Policy"/>
```
Comment (astirpe) [inline account_credit_control/views/res_config_settings_view.xml:17]: Missing `for="..."`.
Or even better, use `<span class="o_form_label">...</span>`

File: account_credit_control/wizard/credit_control_communication.py [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+                                                 group['partner_id'],
+                                                 group['policy_level_id'],
+                                                 group['currency_id']
+                                                 )
+
+            data['credit_control_line_ids'] = [(6, 0, level_lines.ids)]
+            data['partner_id'] = group['partner_id']
+            data['current_policy_level'] = group['policy_level_id']
+            data['currency_id'] = group['currency_id'] or company_currency.id
+            comm = self.create(data)
```
Comment (astirpe) [inline account_credit_control/wizard/credit_control_communication.py:166]: This also can benefit from the batch create, just invoking the create outside the loop.

File: account_credit_control/wizard/credit_control_communication.py [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+                attach_datas = att[1]
+                data_attach = {
+                    'name': attach_fname,
+                    'datas': attach_datas,
+                    'datas_fname': attach_fname,
+                    'res_model': 'mail.mail',
+                    'res_id': email.id,
+                    'type': 'binary',
+                }
+                attachments |= attachments.create(data_attach)
```
Comment (astirpe) [inline account_credit_control/wizard/credit_control_communication.py:212]: This also can probably benefit from the batch create.

File: account_credit_control/wizard/credit_control_policy_changer.py [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+from odoo.exceptions import UserError
+
+
+class CreditControlPolicyChanger(models.TransientModel):
+    """ Wizard that is run from invoices and allows to set manually a policy
+    Policy are actually apply to related move lines availabe
+    in selection widget
+    """
+    _name = "credit.control.policy.changer"
+
```
Comment (astirpe) [inline account_credit_control/wizard/credit_control_policy_changer.py:15]: Missing `_description`

File: account_credit_control/wizard/credit_control_policy_changer.py [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+        :return: ir.actions.act_windows dict
+
+        """
+        self.ensure_one()
+        credit_line_obj = self.env['credit.control.line']
+
+        controlling_date = fields.date.today()
+        self._check_accounts_policies(self.move_line_ids, self.new_policy_id)
+        self._mark_as_overridden(self.move_line_ids)
+        # As disscused with business expert
```
Comment (astirpe) [inline account_credit_control/wizard/credit_control_policy_changer.py:115]: Typo: `disscused` -> `discussed`

**Review by astirpe — CHANGES_REQUESTED** [review]

File: account_credit_control/models/credit_control_line.py (hunk 2 above)
Comment (pedrobaeza) [reply account_credit_control/models/credit_control_line.py:35]: Not exactly, the word order is different.

**Comment by astirpe** [comment]
There is already https://github.com/OCA/credit-control/pull/16 opened

**Comment by pedrobaeza** [comment]
Ouch... that one seems to include already batch create. @cubells can you check that one?

**Comment by cubells** [comment]
Sure!

**Pull request closed without merging**